src/libraries/filcrypto/filproofs/algorithms.go

```diff
@@ -852,66 +862,276 @@ func joinPieceInfos(left PieceInfo, right PieceInfo) PieceInfo {
+	preimage := append(randomness, littleEndianBytesFromUInt(sectorChallengeIndex, 8)...)
+	preimage = append(preimage, littleEndianBytesFromInt(leafChallengeIndex, 8)...)
+	hash := SHA256Hash(preimage)
+	bigHash := bigIntFromLittleEndianBytes(hash)
+
+	challengeSpaceSize := nodes / challengeRangeSize
+	challengeModulus := new(big.Int)
+	challengeModulus.SetUint64(UInt(challengeSpaceSize))
 
-	for _, sector := range challengedSectors {
-		proofAux := sectorStore.GetSectorPersistentProofAux(sector)
-		proofAuxs = append(proofAuxs, proofAux)
+	leafChallenge := bigHash.Mod(bigHash, challengeModulus)
+
+	return leafChallenge.Uint64()
+}
+
+func generateCandidate(randomness sector.PoStRandomness, aux sector.PersistentProofAux, sectorID sector.SectorID, sectorChallengeIndex UInt, leafChallengeCount int, nodes int, challengeRangeSize int) sector.PoStCandidate {
+	treePath := aux.CommRLastTreePath()
+	tree := LoadMerkleTree(treePath)
+
+	var data []byte
+	var inclusionProofs []InclusionProof
+	for i := 0; i < leafChallengeCount; i++ {
+		leafChallenge := generateLeafChallenge(randomness, sectorChallengeIndex, i, nodes, challengeRangeSize)
+
+		for j := 0; j < challengeRangeSize; j++ {
+			leafIndex := leafChallenge + UInt(j)
+			data = append(data, tree.Leaf(leafIndex)...)
+			inclusionProof := tree.ProveInclusion(leafIndex)
+			inclusionProofs = append(inclusionProofs, inclusionProof)
+		}
 	}
 
-	_ = challenges
+	partialTicket := computePartialTicket(randomness, sectorID, data)
 
-	panic("TODO")
+	privateProof := InternalPrivateCandidateProof{
+		InclusionProofs: inclusionProofs,
+	}
+
+	candidate := sector.PoStCandidate_I{
+		PartialTicket_:  partialTicket,
+		PrivateProof_:   privateProof.externalize(),
+		SectorID_:       sectorID,
+		ChallengeIndex_: sectorChallengeIndex,
+	}
+	return &candidate
 }
 
-func (sdr *WinStackedDRG_I) GenerateElectionPoStProof(privateProofs []sector.PrivatePoStProof) sector.PoStProof {
-	panic("TODO")
+func computePartialTicket(randomness sector.PoStRandomness, sectorID sector.SectorID, data []byte) sector.PartialTicket {
+	preimage := randomness
+	preimage = append(preimage, getProverID(sectorID.MinerID())...)
+	preimage = append(preimage, littleEndianBytesFromUInt(UInt(sectorID.Number()), 8)...)
+	preimage = append(preimage, data...)
+	partialTicket := sector.PartialTicket(HashBytes_PedersenHash(preimage))
+
+	return partialTicket
 }
 
-func (sdr *WinStackedDRG_I) VerifyElectionPoSt(sv sector.PoStVerifyInfo) bool {
+func (sdr *WinStackedDRG_I) _generatePoStCandidates(challengeSeed sector.PoStRandomness, eligibleSectors []sector.SectorID, candidateCount int, sectorStore sector_index.SectorStore) (candidates []sector.PoStCandidate) {
+
+	nodes := int(sdr.Nodes())
+	leafChallengeCount := int(sdr.LeafChallengeCount())
+	challengeRangeSize := int(sdr.ChallengeRangeSize())
+	challengedSectors := sdr._getChallengedSectors(eligibleSectors, challengeSeed, eligibleSectors, candidateCount)
```

Comment:
passing in too many args. Does this code compile?

Reply:
It does compile, so I'm not sure why you say 'too many args'.

Reply:
nvm